File: dlna-dms/src/main/java/com/android/cast/dlna/dms/ContentDirectoryService.java
package com.android.cast.dlna.dms;

import org.fourthline.cling.support.contentdirectory.AbstractContentDirectoryService;
import org.fourthline.cling.support.contentdirectory.ContentDirectoryException;
import org.fourthline.cling.support.model.BrowseFlag;
import org.fourthline.cling.support.model.BrowseResult;
import org.fourthline.cling.support.model.SortCriterion;

public class ContentDirectoryService extends AbstractContentDirectoryService {

    @Override
    public BrowseResult browse(String objectID, BrowseFlag browseFlag, String filter,
                               long firstResult, long maxResults, SortCriterion[] orderBy) throws ContentDirectoryException {
        return ContentFactory.getInstance().getContent(objectID);
    }

    @Override
    public BrowseResult search(String containerId, String searchCriteria, String filter,
                               long firstResult, long maxResults, SortCriterion[] orderBy) throws ContentDirectoryException {
        // You can override this method to implement searching!
        return super.search(containerId, searchCriteria, filter, firstResult, maxResults, orderBy);
    }
}

File: dlna-dms/src/main/java/com/android/cast/dlna/dms/ContentDirectoryService.kt
package com.android.cast.dlna.dms

import com.android.cast.dlna.core.Logger
import org.fourthline.cling.support.contentdirectory.AbstractContentDirectoryService
import org.fourthline.cling.support.contentdirectory.ContentDirectoryException
import org.fourthline.cling.support.model.BrowseFlag
import org.fourthline.cling.support.model.BrowseResult
import org.fourthline.cling.support.model.SortCriterion

internal class ContentDirectoryService : AbstractContentDirectoryService() {
    private val logger = Logger.create("ContentDirectoryService")

    @Throws(ContentDirectoryException::class)
    override fun browse(
        objectID: String,
        browseFlag: BrowseFlag,
        filter: String,
        firstResult: Long,
        maxResults: Long,
        orderBy: Array<SortCriterion>,
    ): BrowseResult? {
        logger.i("browse: $objectID, $browseFlag, $filter, $firstResult, $maxResults")
        return ContentFactory.getContent(objectID)
    }

    @Throws(ContentDirectoryException::class)
    override fun search(
        containerId: String,
        searchCriteria: String,
        filter: String,
        firstResult: Long,
        maxResults: Long,
        orderBy: Array<SortCriterion>,
    ): BrowseResult {
        logger.i("search: $containerId, $searchCriteria, $filter, $firstResult, $maxResults")
        return super.search(containerId, searchCriteria, filter, firstResult, maxResults, orderBy)
    }
}